// spinal-model-graph/build/Relations/SpinalRelationFactory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SpinalRelationFactory = exports.RELATION_TYPE_LIST = exports.SPINAL_RELATION_PTR_LST_TYPE = exports.SPINAL_RELATION_LST_PTR_TYPE = exports.SPINAL_RELATION_TYPE = void 0;

var _SpinalRelationRef = _interopRequireDefault(require("./SpinalRelationRef"));

var _SpinalRelationLstPtr = _interopRequireDefault(require("./SpinalRelationLstPtr"));

var _SpinalRelationPtrLst = _interopRequireDefault(require("./SpinalRelationPtrLst"));

var _spinalCoreConnectorjs = _interopRequireDefault(require("spinal-core-connectorjs"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
 * Copyright 2018 SpinalCom - www.spinalcom.com
 *
 * This file is part of SpinalCore.
 *
 * Please read all of the following terms and conditions
 * of the Free Software license Agreement ("Agreement")
 * carefully.
 *
 * This Agreement is a legally binding contract between
 * the Licensee (as defined below) and SpinalCom that
 * sets forth the terms and conditions that govern your
 * use of the Program. By installing and/or using the
 * Program, you agree to abide by all the terms and
 * conditions stated or referenced herein.
 *
 * If you do not agree to abide by these terms and
 * conditions, do not demonstrate your acceptance and do
 * not install or use the Program.
 * You should have received a copy of the license along
 * with this file. If not, see
 * <http://resources.spinalcom.com/licenses.pdf>.
 */
const SPINAL_RELATION_TYPE = "Ref";
exports.SPINAL_RELATION_TYPE = SPINAL_RELATION_TYPE;
const SPINAL_RELATION_LST_PTR_TYPE = "LstPtr";
exports.SPINAL_RELATION_LST_PTR_TYPE = SPINAL_RELATION_LST_PTR_TYPE;
const SPINAL_RELATION_PTR_LST_TYPE = "PtrLst";
exports.SPINAL_RELATION_PTR_LST_TYPE = SPINAL_RELATION_PTR_LST_TYPE;
const RELATION_TYPE_LIST = [SPINAL_RELATION_TYPE, SPINAL_RELATION_LST_PTR_TYPE, SPINAL_RELATION_PTR_LST_TYPE];
exports.RELATION_TYPE_LIST = RELATION_TYPE_LIST;

class SpinalRelationFactory {
  /**
   * Create a new relation of relationType with the relationName.
   * @param {String} relationName Name of the relation
   * @param {String} relationType Type of the relation
   * @return {SpinalRelationRef | SpinalRelationLstPtr | SpinalRelationPtrLst} A new SpinalRelation
   * @static
   */
  static getNewRelation(relationName, relationType) {
    let relation;

    switch (relationType) {
      case SPINAL_RELATION_TYPE:
        relation = new _SpinalRelationRef.default(relationName);
        break;

      case SPINAL_RELATION_LST_PTR_TYPE:
        relation = new _SpinalRelationLstPtr.default(relationName);
        break;

      case SPINAL_RELATION_PTR_LST_TYPE:
        relation = new _SpinalRelationPtrLst.default(relationName);
        break;

      default:
        throw new Error("Unknown relationType");
    }

    return relation;
  }

}

exports.SpinalRelationFactory = SpinalRelationFactory;

_spinalCoreConnectorjs.default.register_models([SpinalRelationFactory]);
